### ISSUE TITLE
stub_const affects object type intermittently

### ISSUE THREAD

**Issue by Jberlinsky** [rspec/rspec-dev#136]
We are observing intermittent failures related to using `stub_const` on objects. It appears that `stub_const`, by some mechanism, is convincing Ruby that an object that was originally a `class` is now a `module`, which breaks when calling `Whatever.new(...)`.

POC follows:

Class definition:

```
class ActivityFeedSerializer < ActiveModel::Serializer
  RESULT_LIMIT = 20

  has_many :items

  def items
    only_valid_items
  end

  private

  def only_valid_items
    ...
  end
end
```

Spec:
```
      it "only returns the first N results" do
        stub_const("ActivityFeedSerializer::RESULT_LIMIT", 2)
        2.times do
          create(:transaction, user: user)
          create(:withdrawal, user: user)
        end

        get endpoint, parameters

        expect(json["activity_feed"]).to have_key "items"
        expect(json["activity_feed"]["items"]).to be_a Array
        expect(json["activity_feed"]["items"].count).to eq 2
      end
    end
```

Every now and then (hard to track down how often, exactly), I'm observing the following:

```
  1) Activity feed retrieval as a logged in user only returns the first N results
     Failure/Error: get endpoint, parameters
     NoMethodError:
       undefined method `new' for ActivityFeedSerializer:Module
```

We can solve this for our particular specs by moving the limit definition to a named method, and stubbing that. However, it is concerning that https://www.relishapp.com/rspec/rspec-mocks/docs/mutating-constants does not make mention of this possible mutation, and leads me to believe that this is a significant bug.

I do not have an opportunity to investigate on a code level at present, but hope to do so shortly.

**Comment by myronmarston** [comment]
It's a side effect of the fact that you aren't explicitly requiring the file that `ActivityFeedSerializer` is defined in.  When `ActivityFeedSerializer` is not already loaded, `stub_const('ActivityFeedSerializer::RESULT_LIMIT', 2)` has no idea whether `ActivityFeedSerializer` is normally a module or a class. The implementation in this case currently defines it as a module since modules are the more limited than classes and it's intended to help with isolation testing so using the more limited form makes sense.   OTOH, when `ActivityFeedSerializer` is already loaded, `stub_const` only messes with the sub-constant `RESULT_LIMIT` and leaves `ActivityFeedSerializer` alone.  The intermittent behavior you are seeing is due to rails auto-loading -- when a spec has run before your `stub_const` call that references `ActivityFeedSerializer`, it auto-loads it so that it is defined, and then it works as you expect.  If such a spec has not already been run, the constant is not defined and RSpec defines it as a module, and you get the error.

The solution is to add `require 'activity_feed_serializer'` (or `require 'models/activity_feed_serializer'` or whatever) to the top of your spec file.  Then it will work consistently.  After all, you depend on that class in that spec file (since you are calling `new` on the class...) so it's a good idea to be sure it is loaded.

In general, Rails autoloading causes problems for testing and is a practice I recommend avoiding.  I've been using explicit requires for many years (and in fact haven't used Rails since 2011).  Autoloading makes things harder to reason about, particularly in a test environment.

I don't think there's any way we can fix this with our current API -- if the constant is not defined we have no way to guess if it is a class or module.  I do think our docs could mention this gotcha, though.  Would you want to open a PR improving our docs here?

**Comment by Jberlinsky** [comment]
More than happy to open a PR improving docs, however, I'm confused as to the usefulness of implicitly defining a module. I might logically expect `stub_const("Foo::BAR", ...)` to raise an error if `Foo` isn't defined. Is there a particular reason for it to implicitly create a `Module`, which makes root-cause analysis of such issues difficult?

**Comment by myronmarston** [comment]
> I might logically expect `stub_const("Foo::BAR", ...)` to raise an error if Foo isn't defined.

`stub_const` came out of rspec-fire, where we wanted to be able to put a `class_double` in place of a class constant, regardless of whether or not the named class is loaded or not.  The goal is to be able to run specs in isolation without dependencies loaded.

I can see why you might logically expect it to work different, but this is how `stub_const` has always been documented to work, and as far as I know, you're the first to express an expectation that it work differently.  Here are the docs from 2.11 (the version in which we added `stub_const`):

https://relishapp.com/rspec/rspec-mocks/v/2-11/docs/stubbing-constants/stub-undefined-constant

> Is there a particular reason for it to implicitly create a `Module`, which makes root-cause analysis of such issues difficult?

To define a nested constant, the parent constants must be defined, and must be either a `Module` or a `Class`.  So the choice is simply whether to create a `Module` or `Class`.  `Module` made more sense to us at the time since modules are the more limited, isolated of the two (classes are subclasses of modules and add additional abilities like instantiation).  Also, when using `stub_const` for `class_double` (e.g. via `class_double("MyGem::SomeClass").as_stubbed_const`), the "normal" thing you'd expect is for `MyGem` to be a module acting as a namespace.  In general, rubyists tend to use modules as namespaces for nested classes much more frequently than classes in my experience.

I'm not convinced that implicitly creating the `Module` makes root cause analysis anymore difficult than creating a `Class` would.  Consider that if we had created a `Class` for you, `ActivityFeedSerializer.new` would have worked but the returned instance would not really have been an instance of the class you expect it to be.  It wouldn't have had any of the methods you expect.  Or in a case where it's `ActivityFeedSerializer.new(:some, :arguments)` it would have failed with a confusing argument error.

Failing fast due to the constant referencing a `Module` instead of `Class` surfaces the problem quicker and, in many cases, would make root cause analysis quicker, I would think.
